MudDataGrid: Configuration using expressions

tjscience opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->
resolves #4567
PR references #5816

This PR is meant to track the changes required to support a property expression as opposed to a string field. The early draft PR is to allow the community and core team to evaluate the progress of this feature as I am developing it out. It is still rough at this point and I have to go through all of the tests, create more and also add better documentation. 

I believe that I have fixed up all of the trimming issues as well.

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] The PR is submitted to the correct branch (`dev`).
- [ ] My code follows the code style of this project.
- [ ] I've added relevant tests.

terry-nextcrew committed:
Majority of work to introduce the new parameter and logic.

terry-nextcrew committed:
progress commit

mikes-gh commented:
@tjscience Thanks for the PR.  Just a note the root of this branch is way back in time and hence the conflicts.  I would advise rebasing this to the head of dev ASAP before doing any more work.

tjscience committed:
Merge branch 'dev' into expression-property

mikes-gh commented:
@tjscience I guess you know but this isn't building.

tjscience commented:
@mikes-gh, it is a rough draft. The whole purpose was to get the foundation out there to be viewed by others. If you want to take a look you should be able to build and run the wasm host project (and obviously every dependency underneath). It is going to take time to go through all of the tests to fix all that this change broke.

mikes-gh commented:
@tjscience check the build errors.  I cant build it :-)

henon commented:
@tjscience: do you have a rough time frame for this PR? I am not pressing you, it is just that as a team we want to plan on how long we are going to have to suppress DataGrid PRs from others. Thanks.

mikes-gh commented:
@tjscience Could you get this building?  Not concerned with the tests for now but it would be good to have it building at least so others could test for you.

tjscience commented:
I am able to build and run the wasm host just fine. Can you give a bit more details about what is failing in the build for you?

![image](https://user-images.githubusercontent.com/3706870/213732595-c8cf40eb-8509-48df-969d-6ec6edb84a80.png)

mikes-gh commented:
The error is in the PR build logs.
These are my local build logs (I only include MudBlazor.dll build to make it shorter)
I am guessing you must be using different code locally.

```
PS /Users/mikes/Documents/Repos/MudBlazor>  cd ./src/MudBlazor
PS /Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor> dotnet build

Welcome to .NET 7.0!
---------------------
SDK Version: 7.0.102

Telemetry
---------
The .NET tools collect usage data in order to help us improve your experience. It is collected by Microsoft and shared with the community. You can opt-out of telemetry by setting the DOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.

Read more about .NET CLI Tools telemetry: https://aka.ms/dotnet-cli-telemetry

----------------
Installed an ASP.NET Core HTTPS development certificate.
To trust the certificate run 'dotnet dev-certs https --trust' (Windows and macOS only).
Learn about HTTPS: https://aka.ms/dotnet-https
----------------
Write your first app: https://aka.ms/dotnet-hello-world
Find out what's new: https://aka.ms/dotnet-whats-new
Explore documentation: https://aka.ms/dotnet-docs
Report issues and find source on GitHub: https://github.com/dotnet/core
Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli
--------------------------------------------------------------------------------------
MSBuild version 17.4.1+9a89d02ff for .NET
  Determining projects to restore...
  Restored /Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj (in 4.1 sec).
  Tool 'excubo.webcompiler' (version '2.7.12') was restored. Available commands: webcompiler
  
  Restore was successful.
  Tool 'excubo.webcompiler' (version '2.7.12') was restored. Available commands: webcompiler
  
  Restore was successful.
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterHeaderCell.razor(52,18): error RZ1034: Found a malformed 'MudMenu' tag helper. Tag helpers must have a start and end tag or be self closing. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(55,32): warning CS8600: Converting null literal or possible null value to non-nullable type. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(69,119): warning CS8604: Possible null reference argument for parameter 'value' in 'bool string.StartsWith(string value, StringComparison comparisonType)'. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(71,117): warning CS8604: Possible null reference argument for parameter 'value' in 'bool string.EndsWith(string value, StringComparison comparisonType)'. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(14,33): warning CS8618: Non-nullable property 'DataGrid' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(18,26): warning CS8618: Non-nullable property 'Column' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(19,23): warning CS8618: Non-nullable property 'Field' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(20,23): warning CS8618: Non-nullable property 'Title' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(21,21): warning CS8618: Non-nullable property 'FieldType' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(22,23): warning CS8618: Non-nullable property 'Operator' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(23,23): warning CS8618: Non-nullable property 'Value' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(24,30): warning CS8618: Non-nullable property 'FilterFunction' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Base/MudFormComponent.cs(587,72): warning IL2075: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperty(String)'. The return value of method 'System.Linq.Expressions.Expression.Type.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Extensions/ExpressionExtensions.cs(41,60): warning IL2075: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperty(String)'. The return value of method 'System.Linq.Expressions.Expression.Type.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterHeaderCell.razor(52,18): error RZ1034: Found a malformed 'MudMenu' tag helper. Tag helpers must have a start and end tag or be self closing. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(55,32): warning CS8600: Converting null literal or possible null value to non-nullable type. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(69,119): warning CS8604: Possible null reference argument for parameter 'value' in 'bool string.StartsWith(string value, StringComparison comparisonType)'. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(71,117): warning CS8604: Possible null reference argument for parameter 'value' in 'bool string.EndsWith(string value, StringComparison comparisonType)'. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(14,33): warning CS8618: Non-nullable property 'DataGrid' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(18,26): warning CS8618: Non-nullable property 'Column' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(19,23): warning CS8618: Non-nullable property 'Field' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(20,23): warning CS8618: Non-nullable property 'Title' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(21,21): warning CS8618: Non-nullable property 'FieldType' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(22,23): warning CS8618: Non-nullable property 'Operator' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(23,23): warning CS8618: Non-nullable property 'Value' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(24,30): warning CS8618: Non-nullable property 'FilterFunction' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Base/MudFormComponent.cs(587,72): warning IL2075: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperty(String)'. The return value of method 'System.Linq.Expressions.Expression.Type.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Extensions/ExpressionExtensions.cs(41,60): warning IL2075: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperty(String)'. The return value of method 'System.Linq.Expressions.Expression.Type.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]

Build FAILED.

/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(55,32): warning CS8600: Converting null literal or possible null value to non-nullable type. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(69,119): warning CS8604: Possible null reference argument for parameter 'value' in 'bool string.StartsWith(string value, StringComparison comparisonType)'. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(71,117): warning CS8604: Possible null reference argument for parameter 'value' in 'bool string.EndsWith(string value, StringComparison comparisonType)'. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(14,33): warning CS8618: Non-nullable property 'DataGrid' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(18,26): warning CS8618: Non-nullable property 'Column' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(19,23): warning CS8618: Non-nullable property 'Field' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(20,23): warning CS8618: Non-nullable property 'Title' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(21,21): warning CS8618: Non-nullable property 'FieldType' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(22,23): warning CS8618: Non-nullable property 'Operator' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(23,23): warning CS8618: Non-nullable property 'Value' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(24,30): warning CS8618: Non-nullable property 'FilterFunction' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Base/MudFormComponent.cs(587,72): warning IL2075: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperty(String)'. The return value of method 'System.Linq.Expressions.Expression.Type.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Extensions/ExpressionExtensions.cs(41,60): warning IL2075: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperty(String)'. The return value of method 'System.Linq.Expressions.Expression.Type.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(55,32): warning CS8600: Converting null literal or possible null value to non-nullable type. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(69,119): warning CS8604: Possible null reference argument for parameter 'value' in 'bool string.StartsWith(string value, StringComparison comparisonType)'. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(71,117): warning CS8604: Possible null reference argument for parameter 'value' in 'bool string.EndsWith(string value, StringComparison comparisonType)'. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(14,33): warning CS8618: Non-nullable property 'DataGrid' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(18,26): warning CS8618: Non-nullable property 'Column' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(19,23): warning CS8618: Non-nullable property 'Field' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(20,23): warning CS8618: Non-nullable property 'Title' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(21,21): warning CS8618: Non-nullable property 'FieldType' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(22,23): warning CS8618: Non-nullable property 'Operator' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(23,23): warning CS8618: Non-nullable property 'Value' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterDefinition.cs(24,30): warning CS8618: Non-nullable property 'FilterFunction' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Base/MudFormComponent.cs(587,72): warning IL2075: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperty(String)'. The return value of method 'System.Linq.Expressions.Expression.Type.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Extensions/ExpressionExtensions.cs(41,60): warning IL2075: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Type.GetProperty(String)'. The return value of method 'System.Linq.Expressions.Expression.Type.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterHeaderCell.razor(52,18): error RZ1034: Found a malformed 'MudMenu' tag helper. Tag helpers must have a start and end tag or be self closing. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net7.0]
/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/Components/DataGrid/FilterHeaderCell.razor(52,18): error RZ1034: Found a malformed 'MudMenu' tag helper. Tag helpers must have a start and end tag or be self closing. [/Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor/MudBlazor.csproj::TargetFramework=net6.0]
    26 Warning(s)
    2 Error(s)

Time Elapsed 00:00:39.21
PS /Users/mikes/Documents/Repos/MudBlazor/src/MudBlazor> 
```

tjscience committed:
Merge branch 'MudBlazor:dev' into expression-property

terry-nextcrew committed:
Make project buildable.

tjscience commented:
@mikes-gh Try it out now and see if you can build.

terry-nextcrew committed:
Fixed some tests.

terry-nextcrew committed:
Fixed issue with filterable in FilterHeaderCell.

terry-nextcrew committed:
Fixed issue with calling add filter from filter panel caused exception.

mikes-gh commented:
> @mikes-gh Try it out now and see if you can build.

Thanks @tjscience I can build now which is great.
When I run the project the css is not working correctly.
Mayb @henon can download an run this PR to see if its just me.

mikes-gh commented:
Ignore previous comment.  After dotnet clean all ok and builds OK @tjscience 👍

jonathan commented:
My team uses MudDataGrid's pretty extensively. Is there a timeline on when this PR might drop?

tjscience committed:
Merge branch 'MudBlazor:dev' into expression-property

terry-nextcrew committed:
Merge branch 'expression-property' of https://github.com/tjscience/MudBlazor into expression-property

terry-nextcrew committed:
Fixed issue where filtering was not taking into consideration an empty value.

terry-nextcrew committed:
Fixed DataGridFilterableTest.
Made filtering expressions more efficient by caching the compilation.

tjscience committed:
Merge branch 'MudBlazor:dev' into expression-property

terry-nextcrew committed:
Fixed some DataGrid tests.
Added extension method to get a column by the property name.
Got rid of the MudDialog component for form editing. Instead, it is now using straight html for the dialog (the issue with double dialogs was the reasoning for this). This is better for perf anyway because it is less nested components.
Fixed issue with nullable column properties and filter definitions.

tjscience committed:
Merge branch 'MudBlazor:dev' into expression-property

terry-nextcrew committed:
Fixed all data grid tests.
Fixed issue with Select/Template columns.

henon commented:
@tjscience I don't want to conflict your PR so if you please address the issue in Filter.cs discussed here: https://github.com/MudBlazor/MudBlazor/issues/6128#issuecomment-1426874443

The time is added to the filter's date but the return value is discarded. In `DateValueChanged` it should be

```c#
          date = date.Add(_valueTime.Value);
```` 

Thanks.

henon commented:
Also this one if you please in `DataGridColumnResizeService.cs` line 52:

![image](https://user-images.githubusercontent.com/44090/218282231-e3d8f6fe-68bf-4a24-9335-a9892ab8e4d2.png)

Taken from: https://pvs-studio.com/en/blog/posts/csharp/1032/

terry-nextcrew committed:
fixes #6128

terry-nextcrew committed:
Fixed issue with DataGridColumnResizeService highlighted in https://pvs-studio.com/en/blog/posts/csharp/1032/

tjscience committed:
Merge branch 'dev' into expression-property

terry-nextcrew committed:
Fixed issue with DataGridFiltersTest.
Fixed auto-generated test issues with PropertyColumn (Had to check if Property parameter was null in OnParametersSet which is not correct in my opinion. This parameter always needs to be set.)

terry-nextcrew committed:
Merge branch 'expression-property' of https://github.com/tjscience/MudBlazor into expression-property

tjscience committed:
Merge branch 'MudBlazor:dev' into expression-property

terry-nextcrew committed:
Cleaned up some files.
Added new example showing the column types.
Added Format property to PropertyColumn.
Added tests.

terry-nextcrew committed:
Fixed format test.

dennisrahmen commented:
You might not have seen the update @henon

henon commented:
@tjscience Thanks for finishing this up. Looking good! 

Is the old way of defining columns with type parameter and string property name completely replaced by the new property column using expressions or can users still use the old method as well? 

How is trimmability of the new property column, I assume better? 

Do you forsee any further redesigns or breaking changes on the component or can we remove the experimental tag now?

henon commented:
@ScarletKuro would you like to add anything with regard to trimming?

tjscience commented:
@henon, this is a true breaking change in that the old way is completely replaced with the new Property way. 

I believe there is more support for trimming with these changes. 

I do not see any other breaking changes going forward. I think we can remove the experimental tag now.

Thanks

ScarletKuro commented:
@henon If its ok i will look on this weekend, since it's a really big change.

ScarletKuro commented:
Trimming wise I haven't found issues. I have many nit comments, there are also some possible null problems, but ig I can express them with a pull request after this is merged?

henon commented:
> Trimming wise I haven't found issues. I have many nit comments, there are also some possible null problems, but ig I can express them with a pull request after this is merged?

Sure

codepend commented:
Is there an ETA when this will be into a release?

dennisrahmen commented:
With the example here: https://dev.mudblazor.com/components/datagrid#grouping
It is already grouped but I can still select "Group" for the column "Molar mass" which results in an error.

tjscience commented:
@dennisrahmen, seems like a invalid cast issue. I will take a look. Thanks.

dennisrahmen commented:
@tjscience should I open new issues or do we track bugs here till the release?

henon commented:
Best to track issues with this PR here.

dennisrahmen commented:
@henon could we get this as a dev nuget update?
That would also help here https://github.com/MudBlazor/MudBlazor/pull/4818#issuecomment-1451633057 and I could test settings I had problems with in the past.

Mr-Technician commented:
This PR's dialog implementation is a bit rough- I have a PR I would like to finish to fix that before the test nuget, if possible. I can do that on my break most likely.

tjscience commented:
What is rough about it exactly? It simplifies the logic, does not depend on components making it more performant, and, most importantly, fixes the issue with dialog that has been around since I created the data grid. A more general note here that I think is important: you should really only use nested components when necessary. It adds more overhead especially in a component library like MudBlazor. I know it is way easier to just throw in MudButtons everywhere, but this does add up in terms of perf hit.

Mr-Technician commented:
The nested dialog issue has been fixed. 😄 I get your point about performance, but without nesting components we would end up repeating ourselves throughout the library. If a change is made to the dialog CSS classes, we now need to manually change those in the DataGrid instead of just modifying the Dialog class.

henon commented:
> @henon could we get this as a dev nuget update? 

Here you go: https://www.nuget.org/packages/MudBlazor/6.1.10-dev1

@Mr-Technician: I can always release another dev pre-release when your's is merged.

henon commented:
@tjscience Mr-Technician fixed the dialog issue in #6390. Are you sure using MudDialog impacts performance of MudDatagrid? Did you measure or is it just an assumption?

henon commented:
> A more general note here that I think is important: you should really only use nested components when necessary. It adds more overhead especially in a component library like MudBlazor.

It is always a tradeoff. You suggest to trade maintainability for performance. If you can make the case that it hits performance significantly I'd consider it, but probably not for marginal gains.

henon commented:
Speaking of performance: we have an interesting PR #6306 waiting to be merged that should make a significant difference to render speed. I don't know how much it will do for DataGrid though

ScarletKuro commented:
I think DataGrid calls this method only once. However, there are other components that call it <5 times.

dennisrahmen commented:
@tjscience is it intentional that the 'TemplateColumn' does allow sorting and filtering if not explicitly disabled? As I thought, it is to be used for action buttons and similar?

Also, a side note, I found that 'GroupExpanded="true"' does not do anything as you can see here the groups are still collapsed on load.
https://try.mudblazor.com/snippet/QYwnanOQLIHlEsHg

tjscience commented:
@dennisrahmen - It was intentional to keep SortBy/GroupBy (filters work a bit differently because we use FilterDefinition for that) on the abstract column instead of only exposing it in the PropertyColumn for a couple of reasons:

1. There are situations where you may want to sort or group by a template column. Maybe your template displays data which includes a combination of properties, or some properties that are not in the data source. You can still do that with this approach. 
2. The Column is abstract and meant to be used to construct other column types as needed. If we do not expose these properties through the column, that functionality would have to be recreated for every inherited column type.

That being said, I think that the template column should sensibly set these parameters if the developer does not explicitly set them, because it doesn't make much sense to show the filter or sort icons for a template column by default (maybe this is closer to what you meant in your comment).

As far as GroupExpanded="true" goes, I can take a look.

lfhenrl commented:
Hi All 
nice work.
I found some bugs.
1. When I use Serverdata with filter the get serverdate is not called when using filtermode ColumnFilterMenu and ColumnFilterRow  It is working with filtermode Simple.
2. When the column type/data is DateTime the filter popup is not showing anything.
3. Every time I open the filter menu/popup It add a new filter definition. My suggest is to only add it when I push the Filter/apply button and the Value is not null.  

Hope it is useful

henon commented:
We need some kind of todo list of regressions that need to be fixed before we can release this into the wild and for all others that can be fixed later we should create issues. At the moment it is hard to keep an overview about what needs to be done. I am making a list below. Maybe I overlooked anything or some are already resolved? If so, let me know.

- [ ] With the example here: https://dev.mudblazor.com/components/datagrid#grouping
It is already grouped but I can still select "Group" for the column "Molar mass" which results in an error. (reported by @dennisrahmen)
- [ ] I found that GroupExpanded="true" does not do anything as you can see here the groups are still collapsed on load. (reported by @dennisrahmen)
- [ ] When I use Serverdata with filter the get serverdate is not called when using filtermode ColumnFilterMenu and ColumnFilterRow It is working with filtermode Simple. (reported by @lfhenrl)
- [ ] When the column type/data is DateTime the filter popup is not showing anything. (reported by @lfhenrl)
- [ ] Every time I open the filter menu/popup It add a new filter definition. My suggest is to only add it when I push the Filter/apply button and the Value is not null. (reported by @lfhenrl)

@tjscience Would you say these are all regressions that need to be fixed before we can release?

tjscience commented:
I will try to go through everything today to see if anything is missing. There is one other item that I would like to get in there but it is not a bug. Being able to set the @key for each row/cell is pretty important and is missing.

Garderoben commented:
@tjscience just a FYI i released this last night as 6.2.0 unaware of this conversation.

henon commented:
PropertyColumn with a nullable datatype causes NullReferenceExceptions with null values (reported by @HClausing). This has been added to [my list of regressions above](https://github.com/MudBlazor/MudBlazor/pull/6049#issuecomment-1467660694)

HClausing commented:
> PropertyColumn with a nullable datatype causes NullReferenceExceptions with null values (reported by @HClausing). This has been added to [my list of regressions above](https://github.com/MudBlazor/MudBlazor/pull/6049#issuecomment-1467660694)

Thanks for add this.

Could it be fixed by #6421 , [by this implementation](https://github.com/MudBlazor/MudBlazor/pull/6421/files#diff-a74d99d3e6ca82e9a4196a76c03bce91b12e096c97b64edbc1ddeab5f78030a9R80) ?

ScarletKuro commented:
> by this implementation

The SetProperty code is identical IL wise.
It's just using pattern matching which is shorter and includes null check

```CSharp
var memberExpression = Property.Body as MemberExpression;
if (memberExpression != null)
```
we can transform to
```CSharp
if (Property.Body as MemberExpression memberExpression)
```
but since below there is another cast `var propertyInfo = memberExpression.Member as PropertyInfo;`
you can go further and use `is { condition }` and make `Property.Body is MemberExpression { Member: PropertyInfo propertyInfo }`

But it's true that in this PR there are other really important null checks added like in the `XYZContext`, `AggregateDefinition`, `FilterDefinition`. And some indeed might help.
But it's hard to say without seeing your code.
If you are using Filters from `AggregateDefinition` then I know what's your problem and how to fix it.
Because indeed if you use nullable value types, lets say for example `int?` and you will have a null in this `int?` field then `AggregateDefinition` will throw on `Sum`, `Average`, `Min`,etc even thought the original LINQ methods do not throw and will accept `nulls`.

henon commented:
> Could it be fixed by #6421 , [by this implementation](https://github.com/MudBlazor/MudBlazor/pull/6421/files#diff-a74d99d3e6ca82e9a4196a76c03bce91b12e096c97b64edbc1ddeab5f78030a9R80) ?

I just merged #6421. I had been reluctant to merge it to avoid conflicts with TJ's further work but I see no PR's from him yet so there seems to be no point in waiting further.

alfadelta88 commented:
Just closed a related issue #6529, NullReferenceExeption is thrown when trying to render a null if the PropertyColumn has formatting. The culprit code is in line 372 of MudDataGrid.razor [here](https://github.com/MudBlazor/MudBlazor/blob/1605ea0e8f2b5f3d712d91ce7ec58558a72c61e3/src/MudBlazor/Components/DataGrid/MudDataGrid.razor#L372). Made a pull request #6541 that should fix it.

alfadelta88 commented:
@tjscience @dennisrahmen `GroupExpanded="true"` works properly when using `ServerData`, I think the problem lies here https://github.com/MudBlazor/MudBlazor/blob/2631eef1a3554400fd240270a89b47a5e890cdcc/src/MudBlazor/Components/DataGrid/MudDataGrid.razor.cs#L1259 the state doesn't get updated when not using ServerData.
Here is a related snippet [link](https://try.mudblazor.com/snippet/mYGdEnRkVXqcSBex)

ScarletKuro commented:
@sjd2021 

> But is it the only way to specify a column now?

There is also TemplateColumn.

> This PR removed the public FieldType and Field on FilterDefinition in favor of, I believe PropertyName and PropertyType on Column<T>.

Do you need them only in FilterDefinition or you want to have them on Column level?

ScarletKuro commented:
Hi @lfhenrl, could you kindly provide a code snippet for the reported issue "When I use Serverdata with filter, the get serverdate is not called when using filtermode ColumnFilterMenu and ColumnFilterRow. It is working with filtermode Simple."? Providing a code snippet will be very helpful in addressing the issue. Thank you.

alfadelta88 commented:
@ScarletKuro are you going to be working on the Serverdata Filter issue, cause I've been meaning to? If so, I'm going to focus on other minor issues with the data grid filtering.

ScarletKuro commented:
Hi @alfadelta88. If you can reproduce the issue or identify its root cause and propose a solution, please feel free to create a PR. I requested the snippet as a fallback option in case no one is available to fix the issue, so that I could potentially take a look and help out. The team is eager to resolve the showstopper issue with the datagrid in order to make a stable release of MudBlazor.

alfadelta88 commented:
@ScarletKuro alright, I'll look into it in the next hours.

ScarletKuro commented:
> After testing it out, it looks like not being able to specify a type will be the only remaining showstopper as I still need DateOnly fields to be treated as DateTime in order to get the calendar picker going

We are aware of the problem with TimeOnly/DateOnly and it will be addressed once the major bugs in the DataGrid are fixed.

> I've got a hacky solution for MudDatePicker to support DateOnly

How does your hacky solution look? `MudDatePicker` supports a `Converter` where you can supply your own converter and map `DateOnly` to `DateTime`, which is a legitimate solution. For this reason, there is no native support for `TimeOnly` in `MudTimePicker` and `DateOnly` in `MudDatePicker` because we genuinely hope that people will use converters.

alfadelta88 commented:
@ScarletKuro as described by @lfhenrl, only simple filtering mode invokes server data reloading, I've also noticed that clearing the filter does not invoke sever data reloading in any mode. I've rushed an initial [commit](https://github.com/MudBlazor/MudBlazor/compare/dev...alfadelta88:MudBlazor:fix/serverdata-filter) with unfinished tests and found out that there is some kind of a bug with enum filtering too (nullable enum maybe?), anyway the server data is reloaded now, maybe you'd like to add some comments before I finish tomorrow.

alfadelta88 commented:
@ScarletKuro as described by @lfhenrl, only simple filtering mode invokes server data reloading, I've also noticed that clearing the filter does not invoke sever data reloading in any mode. I've rushed an initial [commit](https://github.com/alfadelta88/MudBlazor/tree/fix/serverdata-filter) with unfinished tests and found out that there is some kind of a bug with enum filtering too (nullable enum maybe?), anyway the server data is reloaded now, maybe you'd like to add some comments while I finish.

ScarletKuro commented:
@alfadelta88 doesn't the `ReloadServerData` gets triggered twice in `FilterHeaderCell` because `ClearFilter()` it also calling the `void ClearFilter(FilterDefinition<T> filterDefinition)` and they both have the Reload call.

alfadelta88 commented:
@ScarletKuro you're absolutely right, i'll fix it and push it along with the tests in a bit.

ScarletKuro commented:
@alfadelta88 now when #6614 is merged. We can make much better design for you changes.

As all the methods in `FilterHeaderCell` and `HeaderCell` are internal and only used in our handlers I suggest changing the signature of the methods that call `DataGrid.ReloadServerData()` to return a `Task` and add an "-Async" suffix. And then consider using await instead of `.AndForget()`.

Example:
from this
```CSharp
internal void ClearFilter()
{
	...
	if (DataGrid.ServerData != null) DataGrid.ReloadServerData().AndForget();
	...
}
```
to

```CSharp
internal async Task ClearFilterAsync()
{
	...
	if (DataGrid.ServerData != null) await DataGrid.ReloadServerData();
	...
}
```

This will be not a breaking change, because all the methods are internal and only used inside our components, all places that call this methods support returning a Task due to this change #6614.

@henon FYI

alfadelta88 commented:
@ScarletKuro great, I'll make necessary changes.

henon commented:
> @sjd2021 wrote: [...] I'll need to dig into the source code a bit more to see what our options would be for getting the OOB date filter to appear for DateOnly with MudDataGrid. If I come up with something there, is that likely to make it in with the other grid fixes?

Yes, if you want to PR DateOnly support for DataGrid date filter please go ahead. You can tag us in your PR for speedy review. If we release before you are done we can also make a pre-release nuget for you once your PR is in. Just make sure you have good test coverage of your changes.

ScarletKuro commented:
My idea to address this problem is to create an easy abstraction called `ITypeFilterMapper` with an API like this:
```CSharp
public interface ITypeFilterMapper
{
	Type Type { get; }
	RenderFragment RenderFilterFragment<T>(FilterDefinition<T> filterDefinition, Column<T> column);
}
```
This new abstraction can be passed to the DataGrid component. We would define on our side for common mappers for types like String, Boolean, DateTime etc Users can also implement this interface to add support for any custom type they want.
The DataGrid will iterate through the list of `ITypeFilterMapper` and render the necessary fragment.
```CSharp
foreach (ITypeFilterMapper mapper in TypeFilterMapper)
{
	if (filterDefinition.Type == mapper.Type)
	{
		RenderFilterFragment(filterDefinition, filter, column);
	}
}
```
With this approach, users have the flexibility to add any type they want without having to modify the DataGrid component directly. 
Our current solution is that for each type we make a if-clause
```Csharp
@if (filterDefinition.FieldType.IsDateOnly)
{
<HTML CODE>
}
```
And this is not flexible at all.

This is just a high level abstraction that was in my head.

alfadelta88 commented:
@ScarletKuro I pushed a commit to this [branch](https://github.com/alfadelta88/MudBlazor/tree/fix/muddatagrid-serverdata-filter-modes) with the relevant tests, if you could take a look when you have the time. This commit breaks two tests the first is `DataGridColumnPopupCustomFilteringTest` which I'm investigating and the second is `DataGridFilterableServerDataTest` which I think should be invalidated as the filtering should happen at server level when ServerData is defined but the latter expects results with filtering at the client level.

alfadelta88 commented:
@ScarletKuro Btw, with `ColumnFilterRow` the `ReloadServerData` is invoked immediately when a character is inputted in a string filter which is not viable, I think, there should be a public parameter to set `DebounceInterval` to avoid too many server requests, for now I've used a fixed value of 1 second _(in the next commit)_, what do you think?

tjscience commented:
> @ScarletKuro Btw, with `ColumnFilterRow` the `ReloadServerData` is invoked immediately when a character is inputted in a string filter or a number in a numeric filter which is not viable, I think, there should be a public parameter to set `DebounceInterval` to avoid too many server requests, for now I've used a fixed value of 1 second _(̶i̶n̶ ̶t̶h̶e̶ ̶n̶e̶x̶t̶ ̶c̶o̶m̶m̶i̶t̶)̶_, what do you think?

1 second seems like too much of a delay. I think that 500ms would be better to default to.

alfadelta88 commented:
> 1 second seems like too much of a delay. I think that 500ms would be better to default to.

@tjscience You're right, I'll adjust it in the next commit.

neozhu commented:
I am really looking forward to the release of the MudDataGrid component。

henon commented:
@alfadelta88 why don't you create a PR? You can push to it even after you created it, and we can comment and discuss.

ScarletKuro commented:
Yeah, you can make it a draft as well. It will be much easier.

henon commented:
@sjd2021 MudDataGrid has matured, so yes you can work on dateonly and timeonly support without causing conflicts. As for your question, I can't answer it. Maybe @tjscience can.

tjscience commented:
@henon and @sjd2021 The problem with the @if (filterDefinition.FieldType.IsDateOnly) pattern is that it is becoming unwieldy and it increases complexity each time we need to add a type. I am in favor of @ScarletKuro proposed solution [here](https://github.com/MudBlazor/MudBlazor/pull/6049#issuecomment-1502039705). This way all of the types can be managed more easily and overridden by the dev. If this is too much work at this point Kuro, we can go ahead and just implement TimeOnly and DateOnly right now and work on the mapper later.

tjscience commented:
@ScarletKuro, I do not see this implemented cell's edit rendering. Was that a miss, or just not gotten to yet?

ScarletKuro commented:
> @ScarletKuro, I do not see this implemented cell's edit rendering. Was that a miss, or just not gotten to yet?

Not gotten yet, the point was to firstly find an adequate abstraction to dynamically hookup filter parts depending on the type, then everything rest.
